### PR TITLE
Low memory dictionary import POC

### LINKED ISSUE #381: Large Dictionaries Fail to Import on Low Spec Devices
**Description**
On a Pixel 6 running Kiwi Browser with Yomitan, some dictionaries like [Pixiv](https://github.com/MarvNC/yomichan-dictionaries?tab=readme-ov-file#pixiv) fail to import and then crash the page. This was also reported by another user with a chromebook and a user with a 12GB android tablet, as well as a user with an unspecified phone using Firefox for Android.

<details>
  <summary>Image of Crash</summary>

image
</details>

According to Darius,

> Should be possible to fix if someone changes the code to not populate the whole dictionary in RAM before inserting in IndexedDB. Should also increase performance too actually

**Browser version**
Kiwi Browser, Firefox (Android)

**Yomitan version**
Latest

### PR DESCRIPTION
**Pull request by Kuuuube**
Importing Jitendex with this code peaks at roughly 10x less memory usage.

Quick and dirty proof of concept. Will fix #381 and #863.

### CODE REVIEW

**Comment by jamesmaa** [comment]
Would the memory savings be measurable via a playwright test?

**Comment by Kuuuube** [comment]
> Would the memory savings be measurable via a playwright test?

Looks like this exists https://developer.mozilla.org/en-US/docs/Web/API/Performance/measureUserAgentSpecificMemory. Maybe that could be incorporated into the dict import tests. But this pr only helps memory usage for dicts that are split into many files and the current test dicts are probably small enough to be within margin of error. A test that imports a big dict like Jitexdex might be unreasonable.

**Comment by Kuuuube** [comment]
This should be good to go now.

Some tests I did with this using Jitendex 2025-08-01 on Chromium. None of these tests are super perfect but I think it does a decent job at showing that this PR saves a lot of memory usage and doesn't cause slowdowns in importing.

Jitendex alone doesn't cover every possible edge case for dict imports but it uses quite a lot so a Jitendex import with no difference in the database shows quite well that it shouldn't cause issues with importing for almost all dictionaries.

I encourage anyone to give this a test with your dictionaries and check if you see any problems.

1. Jitendex dict collection diff, no diff between master and this pr when exporting the database.
    It is rather hard to diff json files this big (it crashes `git diff`) and I had to make some basic custom tooling, if anyone wants to run this test themselves on any dictionaries let me know.
3. Import speed on Jitendex:
    Testing steps:
    1. Restart Yomitan from the extensions page
    2. Open the settings page
    4. Open the dict importer
    5. Import the dictionary file and start the timer
    6. End the timer as soon as the progress bar disappears from the UI

    | Master | This PR |
    |:-|:-|
    | 0:50 | 0:39 |
    | 0:39 | 0:33 |
    | 0:44 | 0:43 |
3. Memory usage on Jitendex:
    Testing steps:
    1. Restart Yomitan from the extensions page
    2. Open the settings page
    3. Open the dict importer
    7. Record `Before Import` memory usage
    8. Select the dictionary file and import the dictionary
    9. Watch and record `Peak During Import` memory
    10. After import finishes wait a few seconds for Chromium's GC to run and record the `After Import` memory

    | Test 1 | Master | This PR |
    |:-|:-|:-|
    | Before Import | 1.5gb | 1.5gb |
    | Peak During Import | 2.6gb | 1.6gb |
    | After Import (after GC runs) | 1.5gb | 1.5gb |
    | Estimated Used by Yomitan Import | 1.1gb | 0.1gb |

    | Test 2 | Master | This PR |
    |:-|:-|:-|
    | Before Import | 1.5gb | 1.5gb |
    | Peak During Import | 2.8gb | 1.7gb |
    | After Import (after GC runs) | 1.5gb | 1.5gb |
    | Estimated Used by Yomitan Import | 1.3gb | 0.2gb |

    | Test 3 | Master | This PR |
    |:-|:-|:-|
    | Before Import | 1.5gb | 1.5gb |
    | Peak During Import | 2.8gb | 1.7gb |
    | After Import (after GC runs) | 1.5gb | 1.5gb |
    | Estimated Used by Yomitan Import | 1.3gb | 0.2gb |

**Comment by Kuuuube** [comment]
Due to the way this imports, validation happens incrementally. This can allow an invalid dictionary to partially write to the database. For example: if term banks 1-50 are valid but term bank 51 is invalid, 1-50 will be written then it will error on 51 and stop. Invalid data will still never be written but bad dicts could partially import.

The only way to 100% avoid this without ballooning memory usage would require loading the file data twice which gives some hit to speed. Instead I've added a marker to the dictionary that will show whether it imported successfully or not. This will always show when the user opens their dictionaries and is not a database check (like the check integrity button).

image

Edit: Added in validation before import for now. [5fa854e](https://github.com/yomidevs/yomitan/pull/2001/commits/5fa854e5607806a750ce8f6a2795f28a204be5e0)

**Comment by Kuuuube** [comment]
New benchmarks on Jitendex after the validation changes. Running this on a system that imports roughly 3x slower to get more consistent numbers.

| Test 1 | Master | This PR |
|:-|:-|:-|
| Before Import | 383mb | 403mb |
| Peak During Import | 1780mb | 675mb |
| After Import (after GC runs) | 422mb | 414mb |
| Estimated Used by Yomitan Import | 1377mb | 266mb |
| Time Elapsed | 1:21 | 1:25 |

| Test 2 | Master | This PR |
|:-|:-|:-|
| Before Import | 450mb | 452mb |
| Peak During Import | 1699mb | 751mb |
| After Import (after GC runs) | 457mb | 434mb |
| Estimated Used by Yomitan Import | 1245mb | 308mb |
| Time Elapsed | 1:30 | 1:26 |

| Test 3 | Master | This PR |
|:-|:-|:-|
| Before Import | 436mb | 444mb |
| Peak During Import | 1785mb | 681mb |
| After Import (after GC runs) | 435mb | 434mb |
| Estimated Used by Yomitan Import | 1349mb | 242mb |
| Time Elapsed | 1:39 | 1:25 |

Seems like speed has not taken a hit here so this is the way to go (memory is fine too, no differences were expected with these changes).

File: ext/js/data/database.js [hunk 1]
```diff
@@ -135,6 +135,55 @@ export class Database {
...
+
+            if (count <= 0) {
+                resolve();
+                return;
+            }
+
+            const transaction = this._readWriteTransaction([objectStoreName], resolve, reject);
+            const objectStore = transaction.objectStore(objectStoreName);
+
+            for (let i = start, ii = start + count; i < ii; ++i) {
```
Comment (MarvNC) [inline ext/js/data/database.js:180]: Why not just do
```js
            for (let i = start; i < start + count; ++i) {
```

File: ext/js/dictionary/dictionary-importer.js [hunk 2]
```diff
@@ -67,6 +67,37 @@ export class DictionaryImporter {
...
+        const maxTransactionLength = 1000;
+        const bulkAddProgressAllowance = 1000;
+
+        /**
+         * @template {import('dictionary-database').ObjectStoreName} T
+         * @param {T} objectStoreName
+         * @param {import('dictionary-database').ObjectStoreData<T>[]} entries
+         */
+        const bulkAdd = async (objectStoreName, entries) => {
+            const ii = entries.length;
```
Comment (MarvNC) [inline ext/js/dictionary/dictionary-importer.js:81]: what does ii mean

**Review by MarvNC — APPROVED** [review]

File: ext/js/dictionary/dictionary-importer.js (hunk 2 above)
Comment (Kuuuube) [reply ext/js/dictionary/dictionary-importer.js:81]: Also not a line I wrote though it appears in my diff because it got moved around. Will rename to `entryCount`.

**Review by MarvNC — APPROVED** [review]
